src/lib/gssapi/krb5/init_sec_context.c

```diff
@@ -582,8 +587,18 @@ kg_new_connection(
                                   &ctx->there)))
         goto cleanup;
 
-    code = get_credentials(context, cred, ctx->there, now,
-                           ctx->krb_times.endtime, &k_cred);
+    if (cred->impersonator) {
```

Comment:
If instead of `probe` you pass in `flags`, then this chunk of code could have just two calls to `get_credentials()`.  You could have a local `flags` initialized to 0 then:

    if (cred->impersonator)
        flags |= KRB5_GC_CACHED;
    code = get_credentials(context, cred, ctx->there, now, ctx->krb_times.endtime, &k_cred, flags);
    if (code == KRB5_CC_NOTFOUND) {
        flags &= ~KRB5_GC_CACHED;
        code = get_credentials(context, cred, ctx->there, now, ctx->krb_times.endtime, &k_cred, flags);
    }
    if (code)
        goto cleanup;